HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -0,0 +1,129 @@
+            cancellationTokenSource.CancelAfter(TimeSpan.FromMinutes(5));
+
+            // get the JSON data
+            var resultJson = await _downloaderService.DownloadStringAsync(VmGalleryUrl.AbsoluteUri, cancellationTokenSource.Token);
+            _imageList = JsonSerializer.Deserialize(resultJson, typeof(VMGalleryImageList), _jsonOptions) as VMGalleryImageList ?? emptyList;
+
+            // Now we need to download the base64 images for the symbols (icons). So they can be used within an adaptive card.
+            foreach (var image in _imageList.Images)
+            {
+                if (!string.IsNullOrEmpty(image.Symbol.Uri))
+                {
+                    var byteArray = await _downloaderService.DownloadByteArrayAsync(image.Symbol.Uri, cancellationTokenSource.Token);
+                    if (!ValidateFileSha256Hash(byteArray, image.Symbol.Hash))
+                    {
+                        Logging.Logger()?.ReportError(ComponentName, $"Symbol Hash '{image.Symbol.Hash}' validation failed for image with name '{image}'. Symbol uri: '{image.Symbol}'");
+                        continue;
+                    }
+
+                    image.Symbol.Base64Image = Convert.ToBase64String(byteArray);
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Logging.Logger()?.ReportError(ComponentName, $"Unable to retrieve VM gallery images", ex);
+        }
+
+        return _imageList;
+    }
+
+    /// <inheritdoc cref="IVMGalleryService.GetDownloadedArchiveFileName"/>
+    public string GetDownloadedArchiveFileName(VMGalleryImage image)
+    {
+        var hash = image.Disk.Hash;
+        var hexValues = hash.Split(':').Last();
+
+        // Now get the file extension from the web uri. The web uri is the location of the zipped image on the web.
+        // we can expect this last segment to be the file name plus the extension.
+        var webUri = new Uri(image.Disk.Uri);
+        var fileExtension = webUri.Segments.Last().Split('.').Last();
+        return $"{hexValues}.{fileExtension}";
+    }
+
+    /// <summary>
+    /// Validates that the file has the correct SHA256 hash and has not been tampered with.
+    /// </summary>
+    /// <param name="file">File on the file system that we will compare the hash to</param>
+    /// <returns>True if the hash validation was successful and false otherwise</returns>
+    public async Task<bool> ValidateFileSha256Hash(StorageFile file)
+    {
+        var fileStream = await file.OpenStreamForReadAsync();
+        var hashedFileStream = SHA256.HashData(fileStream);
+        var hashedFileString = BitConverter.ToString(hashedFileStream).Replace("-", string.Empty);
+
+        // For our usage the file name before the extension is the original hash of the file.
+        // We'll compare that to the current hash of the file.
+        return string.Equals(hashedFileString, file.Name.Split('.').First(), StringComparison.OrdinalIgnoreCase);
```

Comment:
`.first()` can throw if `file.Name` is `null`.  Suggest use `FiratOrDefault()` and null-coalesc to string.empty.

Reply:
Its ok for it to throw, I expect the caller to catch exceptions from this method.